### PR TITLE
Add trace resolver / data loader

### PR DESCRIPTION
**Pull request by darwin67**
## Description

This change add GQL resolver and data loader for trace spans that power the function run detail data for the new UI.
It covers the following ops
- StepRun (including grouping for retries)
- StepSleep
- StepWaitForEvent
- StepInvoke

~80% of the API is working as expected.
The rest if mainly adding more test coverage to make sure edge cases show correctly, but the change can be reviewed and merged as is.

### Follow ups

What likely need further testing:
- [X] Function retries
- [ ] Edge cases and failure scenarios (wait/invoke timeout, cancellation, etc)
- [x] Generalize the span tree construction logic. 

## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [x] I've linked any associated issues to this PR.
- [x] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/coreapi/graph/loaders/trace.go [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+	}
+	if ttb.trigger == nil {
+		return nil, fmt.Errorf("no trigger span found")
+	}
+
+	return ttb, nil
+}
+
+// Build goes through the tree and construct the trace for API to be consumed
+func (tb *TraceTreeBuilder) Build(ctx context.Context) (*models.RunTraceSpan, error) {
```
Comment (darwin67) [inline pkg/coreapi/graph/loaders/trace.go:283]: This is where the core of where the tree is built

File: pkg/coreapi/graph/loaders/trace.go [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+			status = models.RunTraceSpanStatusFailed
+		}
+		mod.Status = status
+	} else {
+		if timedOut != nil && *timedOut {
+			mod.Status = models.RunTraceSpanStatusFailed
+		}
+	}
+
+	// TODO: output
```
Comment (jpwilliams) [inline pkg/coreapi/graph/loaders/trace.go:715]: Later PR?

**Review by jpwilliams — APPROVED** [review]
🙌

File: pkg/devserver/span.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func (sh *spanIngestionHandler) Add(ctx context.Context, span *cqrs.Span) {
 	_, _ = h.Write([]byte(id))
 	key := base64.StdEncoding.EncodeToString(h.Sum(nil))
 
-	if s, ok := sh.dedup[key]; !ok || span.Duration >= s.Duration {
+	if s, ok := sh.dedup[key]; !ok || (ok && span.Duration >= s.Duration) {
```
[the remark this answers is not in the document]
Comment (darwin67) [reply pkg/devserver/span.go:58]: Not really, they're all basically identical except the duration, and maybe the nanosecond level at timestamp, but millisecond precision is always the same.

File: pkg/coreapi/graph/loaders/trace.go (hunk 2 above)
Comment (darwin67) [reply pkg/coreapi/graph/loaders/trace.go:715]: Yup, output and trigger APIs are separate PRs. this PR itself is getting big.
